src/sinks/statsd.rs

```diff
@@ -206,30 +205,47 @@ fn encode_event(event: Event, namespace: &str) -> Option<Vec<u8>> {
 
 impl Service<Vec<u8>> for StatsdSvc {
     type Response = ();
-    type Error = tokio::io::Error;
-    type Future = future::Ready<Result<(), Self::Error>>;
+    type Error = StatsdError;
+    type Future = future::BoxFuture<'static, Result<(), Self::Error>>;
 
     fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
         Poll::Ready(Ok(()))
     }
 
     fn call(&mut self, mut frame: Vec<u8>) -> Self::Future {
-        // remove trailing delimiter
+        use futures::compat::Sink01CompatExt;
+        use futures::Sink;
+        use futures::SinkExt;
+        use std::pin::Pin;
+
+        type ByteSink = Pin<Box<dyn Sink<bytes::Bytes, Error = ()> + 'static + Send>>;
+
         if let Some(b'\n') = frame.last() {
             frame.pop();
-        };
-        self.client.send(frame.as_ref());
-        future::ok(())
+        }
+        let client = self.client.clone();
+        async move {
+            let mut sink: ByteSink = match client {
+                Client::Udp(inner) => Box::pin(inner.into_sink().context(UdpError)?.sink_compat()),
+                Client::Tcp(inner) => Box::pin(inner.into_sink().sink_compat()),
+                #[cfg(unix)]
+                Client::Unix(inner) => Box::pin(inner.into_sink().sink_compat()),
+            };
+            sink.send(frame.into())
+                .await
+                .map_err(|_| StatsdError::SendError)?;
```

Comment:
`tcp`, `udp`, and `unix` sinks use `()` as error type, so there is no error source here.